trace-normalization/src/normalizer.rs

```diff
@@ -103,11 +113,71 @@ pub(crate) fn is_valid_status_code(sc: &str) -> bool {
     false
 }
 
+// normalize_trace takes a trace and
+// * returns an error if there is a trace ID discrepancy between 2 spans
+// * returns an error if at least one span cannot be normalized
+pub fn normalize_trace(trace: &mut [pb::Span]) -> anyhow::Result<()> {
+    if trace.is_empty() {
+        anyhow::bail!("Normalize Trace Error: Trace is empty.");
+    }
+
+    let first_trace_id = trace[0].trace_id;
+
+    for span in trace {
+        if span.trace_id != first_trace_id {
+            anyhow::bail!(format!(
+                "Normalize Trace Error: Trace has foreign span: {:?}",
+                span
+            ));
+        }
+        match normalize(span) {
+            Ok(_) => {}
+            Err(e) => anyhow::bail!(e),
+        }
+    }
+    Ok(())
+}
+
+// normalize_chunk takes a trace chunk and
+// * populates origin field if it wasn't populated
+// * populates priority field if it wasn't populated
+// the root span is used to populate these fields, and it's index in TraceChunk spans vec must be passed.
+pub fn normalize_chunk(chunk: &mut pb::TraceChunk, root_span_index: usize) -> anyhow::Result<()> {
+    anyhow::ensure!(
+        chunk.spans.len() > root_span_index,
+        "Normalize Chunk Error: root_span_index > length of trace chunk spans"
+    );
+    // check if priority is not populated
+    let root_span = &chunk.spans[root_span_index];
+    if chunk.priority == SAMPLER_PRIORITY_NONE {
+        // Older tracers set sampling priority in the root span.
+        if let Some(root_span_priority) = root_span.metrics.get(TAG_SAMPLING_PRIORITY) {
+            chunk.priority = *root_span_priority as i32;
+        } else {
+            for span in &chunk.spans {
```

Comment:
I'm curious why normalize_chunk() doesn't call normalize() on this span list anywhere. Do we do that in another function somewhere?

Reply:
Sorry,  I think I meant normalize_chunk calling normalize_trace

Reply:
ah, I see. `normalize_trace` is called independently from `normalize_chunk`, who's sole purpose is just to populate the `origin` and `priority` fields. [See here in agent.to where the normalizer is called](https://github.com/DataDog/datadog-agent/blob/db20bdd7f29132a243c28feb178a5ff6514aaa90/pkg/trace/agent/agent.go#L252)